### PR TITLE
[Serve] Define Java Backend

### LINKED ISSUE #15835: [serve] Define Java backend core API
**Design Document** : https://docs.google.com/document/d/1b9VCZsC2jVhWw_c_ifMu_ED3NxkARcOnU-OFqUYu-r4/edit#heading=h.b3vvz7iwnixn

### PR DESCRIPTION
**Pull request by liuyang-my**
## Why are these changes needed?

We hope to take Ray Serve to AntGroup and reuse the ability of Ray Serve in Ant's Ray Serving. In Ant, a large part of business scenarios come from Java developers, so Ray Serve is required to support Java backend.

In the milestone "[serve] Support Java Backend", we will enable Ray Serve to support Java Backend. Detailed design is in doc :https://docs.google.com/document/d/1b9VCZsC2jVhWw_c_ifMu_ED3NxkARcOnU-OFqUYu-r4/edit#heading=h.b3vvz7iwnixn

This PR is about the definition of Java Backend.

## Related issue number

#15835 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by liuyang-my** [comment]
@jovany-wang

File: java/serve/src/main/java/io/ray/serve/BackendConfig.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package io.ray.serve;
+
+import com.google.common.base.Preconditions;
+import java.io.Serializable;
+import org.apache.commons.lang3.builder.ReflectionToStringBuilder;
+
+/**
```
Comment (simon-mo) [inline java/serve/src/main/java/io/ray/serve/BackendConfig.java:7]: Let's make sure to convert this into protobuf in the next PR?

File: java/serve/src/main/java/io/ray/serve/poll/LongPollClient.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  /** An async thread to post the callback into. */
+  private Thread pollThread;
+
+  public LongPollClient(BaseActorHandle hostActor, Map<KeyType, KeyListener> keyListeners) {
+
+    Preconditions.checkArgument(keyListeners != null && keyListeners.size() != 0);
+
+    this.hostActor = hostActor;
+    this.keyListeners = keyListeners;
+    this.snapshotIds = new ConcurrentHashMap<>();
```
Comment (simon-mo) [inline java/serve/src/main/java/io/ray/serve/poll/LongPollClient.java:44]: 👍

**Review by simon-mo — APPROVED** [review]
This looks great! I think the next step here that we need to start converting majority of the data classes (POJO classes) into protobuf types.

**Comment by simon-mo** [comment]
I think there are several format and lint issues, 
https://buildkite.com/ray-project/ray-builders-pr/builds/6316#ce636c2f-ec20-4069-b28f-613bfc5124c0/315-368
https://buildkite.com/ray-project/ray-builders-pr/builds/6316#500bbdb9-aa28-432f-929f-19c17616e555/140-375

Additionally, are these tests ran at all? Can you try add them to https://github.com/ray-project/ray/blob/master/java/test.sh

File: java/serve/src/main/java/io/ray/serve/BackendConfig.java (hunk 1 above)
Comment (liuyang-my) [reply java/serve/src/main/java/io/ray/serve/BackendConfig.java]: Yes. I'll design some plans about next PR this week.

**Comment by simon-mo** [comment]
@liuyang-my ping about adding the tests. Thanks!

**Review by caitengwei — APPROVED** [review]
LGTM

File: java/serve/pom.xml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    </dependency>
+    <dependency>
+      <groupId>io.ray</groupId>
+      <artifactId>ray-runtime</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+<dependency>
+  <groupId>com.google.guava</groupId>
+  <artifactId>guava</artifactId>
+  <version>27.0.1-jre</version>
```
Comment (jovany-wang) [inline java/serve/pom.xml:33]: Unfortunately, we had upgraded guava to 30.0 to fix CVEs. The PR is https://github.com/ray-project/ray/pull/16650

Comment (jovany-wang) [reply java/serve/pom.xml:33]: @liuyang-my Can you also fix it?

**Pull request merged**